mission_control/src/mission_control.cpp

```diff
@@ -271,91 +194,94 @@ void MissionControlNode::loadMissionCallback(const mission_control::LoadMission:
 }
 
 void MissionControlNode::executeMissionCallback(
     const mission_control::ExecuteMission::ConstPtr& msg)
 {
-  // TODO(qna): Need to shutdown any mission that is currently running.
-  ROS_INFO("executeMissionCallback - Executing mission id[%d]", msg->mission_id);
-  executeMission(msg->mission_id);
+  if (m_current_mission_id > 0)
+  {
+    BT::NodeStatus missionStatus = m_mission_map[m_current_mission_id]->getMissionStatus();
+    if ((missionStatus != BT::NodeStatus::IDLE) && (missionStatus != BT::NodeStatus::SUCCESS))
+    {
+      ROS_INFO_STREAM("There is a mission being executed - mission id["
+                      << m_current_mission_id << "] - Mission Status: "
+                      << m_mission_map[m_current_mission_id]->getMissionStatus());
+      return;
+    }
+  }
+
+  if ((msg->mission_id > 0) && (m_mission_map.size() > 0) &&
+      (m_mission_map.count(msg->mission_id) > 0))
+  {
+    m_current_mission_id = msg->mission_id;
+    executeMissionTimer.start();
+    ROS_INFO_STREAM("executeMissionCallback - Executing mission id[" << m_current_mission_id
+                                                                     << "]");
+  }
+  else
+  {
+    ROS_INFO_STREAM("Error - Wrong Mission Number. "
+                    << "Mission Number: " << msg->mission_id);
+  }
 }
 
 void MissionControlNode::abortMissionCallback(const mission_control::AbortMission::ConstPtr& msg)
 {
-  // TODO(qna): Need to shutdown any mission that is currently running.
-  // TODO(qna): should probably also check the timestamp from the message
-  ROS_INFO("abortMissionCallback - Aborting mission id[%d]", msg->mission_id);
-  abortMission(msg->mission_id);  // assume mission ID is the current mission
+  abortMission(msg->mission_id);
+  ROS_INFO_STREAM("abortMissionCallback - Aborting mission id " << msg->mission_id);
 }
 
 void MissionControlNode::queryMissionsCallback(const mission_control::QueryMissions::ConstPtr& msg)
 {
   ReportMissions outmsg;
   MissionData data;
-  std::map<int, Mission*>::iterator it;
-
-  for (it = m_mission_map.begin(); it != m_mission_map.end(); it++)
+  for (const auto& it : m_mission_map)
   {
-    data.mission_id = it->first;
-    data.mission_description = it->second->getMissionDescription();
+    data.mission_id = it.first;
+    data.mission_description = it.second->getCurrentMissionDescription();
     outmsg.missions.push_back(data);
   }
 
   outmsg.header.stamp = ros::Time::now();
   pub_report_missions.publish(outmsg);
 
-  ROS_INFO("queryMissionsCallback - Reporting these missions");
-  for (int i = 0; i < outmsg.missions.size(); ++i)
+  ROS_INFO_STREAM("queryMissionsCallback - Reporting these missions");
+  for (const auto& mission : outmsg.missions)
   {
-    const mission_control::MissionData& mdata = outmsg.missions[i];
-    ROS_INFO("Mission Id:[%d] Description:[%s]", mdata.mission_id,
-             mdata.mission_description.c_str());
+    const mission_control::MissionData& mdata = mission;
```

Comment:
@patapufate nit: is this local reference even needed? Why not using `mission` where `mdata` is used?

Reply:
Fixed